vindex/cmd/logandmap/main.go

```diff
@@ -0,0 +1,331 @@
+			} else {
+				klog.Infof("Appended entry for %s@%s at index %d", module, version, idx.Index)
+			}
+		}
+	}
+}
+
+func runWebServer(vi *vindex.VerifiableIndex) {
+	web := NewServer(func(s string) string {
+		idxes := vi.Lookup(s)
+		return fmt.Sprintf("Indices in log: %v", idxes)
+	})
+
+	fs := http.FileServer(http.Dir(*posixLogDir))
+	r := mux.NewRouter()
+	r.PathPrefix("/log/").Handler(http.StripPrefix("/log/", fs))
+	web.registerHandlers(r)
+	hServer := &http.Server{
+		Addr:    *addr,
+		Handler: r,
+	}
+	go func() {
+		_ = hServer.ListenAndServe()
+	}()
+	klog.Infof("Started HTTP server listening on %s", *addr)
+}
+
+// Read log private key from file or environment variable and generate the
+// note Signer and Verifier pair for it.
+func getSignerVerifierOrDie() (note.Signer, note.Verifier) {
+	var privKey string
+	var err error
+	if len(*privKeyFile) > 0 {
+		privKey, err = getKeyFile(*privKeyFile)
+		if err != nil {
+			klog.Exitf("Unable to get private key: %v", err)
+		}
+	} else {
+		privKey = os.Getenv("LOG_PRIVATE_KEY")
+		if len(privKey) == 0 {
+			klog.Exit("Supply private key file path using --private_key or set LOG_PRIVATE_KEY environment variable")
+		}
+	}
+	s, v, err := signerVerifierFromSkey(privKey)
+	if err != nil {
+		klog.Exitf("Failed to get signer/verifier: %v", err)
+	}
+	return s, v
+}
+
+func signerVerifierFromSkey(skey string) (note.Signer, note.Verifier, error) {
+	const algEd25519 = 1
+	s, err := note.NewSigner(skey)
+	if err != nil {
+		return nil, nil, err
+	}
+	_, skey, _ = strings.Cut(skey, "+")
+	_, skey, _ = strings.Cut(skey, "+")
+	_, skey, _ = strings.Cut(skey, "+")
+	_, key64, _ := strings.Cut(skey, "+")
+	key, err := base64.StdEncoding.DecodeString(key64)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode base64: %v", err)
+	}
+
+	alg, key := key[0], key[1:]
+	switch alg {
+	default:
+		return nil, nil, errors.New("unsupported algorithm")
+
+	case algEd25519:
+		if len(key) != ed25519.SeedSize {
+			return nil, nil, fmt.Errorf("expected key seed of size %d but got %d", ed25519.SeedSize, len(key))
+		}
+		key := ed25519.NewKeyFromSeed(key)
+		publicKey := key.Public().(ed25519.PublicKey)
+		vkey, err := note.NewEd25519VerifierKey(s.Name(), publicKey)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to generate verifier from key: %v", err)
+
+		}
+		v, err := note.NewVerifier(vkey)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create verifier from vkey: %v", err)
+		}
+		return s, v, err
+	}
+}
+
+func getKeyFile(path string) (string, error) {
+	k, err := os.ReadFile(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to read key file: %w", err)
+	}
+	return string(k), nil
+}
+
+func mapFnFromFlags() vindex.MapFn {
+	mapFn := func(data []byte) [][32]byte {
+		var entry LogEntry
+		if err := json.Unmarshal(data, &entry); err != nil {
+			panic(fmt.Errorf("failed to unmarshal entry: %v", err))
+		}
+
+		return [][32]byte{sha256.Sum256([]byte(entry.Module))}
```

Comment:
Just a thought: I wonder if there's value in having a preimage here which mixes in _something_ about the semantics of the mapping? e.g. `SHA256("mapping: by module\ndata: <data>")` (or something) - the (half-baked) thought was two-fold: one: this makes it explicit to the believer what it is they're looking up "by", and two: this would enable the same map instance to answer multiple "questions" about the data.

Reply:
Added a comment about this possibility but will avoid code changes for now. There are competing tensions here, as Rasmus and Filippo would like this key function to return a preimage instead of a hash, which could allow prefix searching in a trie.